Use live illink to trim framework

sbomer opened the pull request:
https://github.com/dotnet/runtime/pull/90517 added a workaround to trim the framework using the SDK's version of ILLink.Tasks (see https://github.com/dotnet/runtime/pull/90517#issuecomment-1677995252 for context). This removes the workaround and replaces it with logic to use the live build of ILLink.Tasks.

The dependency on ILLink.Tasks is expressed via a magical `ProjectReference` that @ViktorHofer kindly helped me construct.

Unfortunately I had to add a small extension point to `Microsoft.NET.ILLink.Tasks.props`, to allow overriding the analyzer props path. This lets it be imported right out of the source tree, instead of from the build output of ILLink.Tasks.

sbomer committed:
Use live illink

sbomer committed:
Include ILLink.Tasks for subsets that depend on it

Also prevent build-native.proj from importing illink.targets.

ViktorHofer commented on eng/Subsets.props:
```diff
@@ -453,6 +455,7 @@
   </ItemGroup>
 
   <ItemGroup Condition="$(_subset.Contains('+libs.ref+')) or $(_subset.Contains('+libs.src+')) or $(_subset.Contains('+libs.sfx+'))">
+    <ProjectToBuild Include="$(ToolsProjectRoot)illink\src\ILLink.Tasks\ILLink.Tasks.csproj" />
```
Doing this means that we can't do a `dotnet build src\libraries\System.Text.Json\src\ -f net7.0` (any target framework, net7.0 is just an example) without first building via the build command script. That's because you use the Subsets system here to define dependencies, which only works for the root build.

Instead, we should add a ProjectReference to ILLink.Tasks.csproj to the projects that need it. Presumably that are all the projects that import illink.targets. Consider adding a ProjectReference to illink.targets and remove these entries here.

ViktorHofer replied:
> I was debating the best way to express the dependency. One problem is that ILLink.Tasks.csproj builds an MSBuild task, so I don't think building it as a ProjectReference would make the task available early enough for a UsingTask in the consuming project.

ProjectReferences are built as part of the `ResolveProjectReferences` target which runs before the CoreCompile target in which ILLink is invoked. The UsingTask only needs the path to the assembly at evaluation time, but the assembly doesn't need to exist at evaluation time.

The task subset is just another abomination which we eventually need to fix and declare dependencies in the consuming projects. I would recommend to ignore that.

ViktorHofer commented on eng/illink.targets:
```diff
@@ -1,5 +1,7 @@
 <Project>
 
+  <Import Project="$(ToolsILLinkDir)$(NetCoreAppToolCurrent)/build/Microsoft.NET.ILLink.Tasks.props" />
```
It's important to only import files that exist at msbuild evaluation time. Where does that file come from?

ViktorHofer replied:
Can you import it from the checked-in place instead? https://github.com/dotnet/runtime/blob/main/src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.Tasks.props. Btw it's a weird for a targets file to import a props file. We might want to clean that up eventually when consolidating the stuff in illink.targets with Microsoft.NET.ILLink.Tasks.targets.

ViktorHofer commented on src/libraries/Directory.Build.targets:
```diff
@@ -127,7 +127,7 @@
   <Import Project="$(RepositoryEngineeringDir)slngen.targets" Condition="'$(IsSlnGen)' == 'true'" />
 
-  <Import Project="$(RepositoryEngineeringDir)illink.targets" Condition="'$(IsSourceProject)' == 'true'" />
+  <Import Project="$(RepositoryEngineeringDir)illink.targets" Condition="'$(IsSourceProject)' == 'true' and '$(MSBuildProjectExtension)' != '.proj'" />
```
Do you remember which projects needed this condition? Just curious.

ViktorHofer replied:
Hmm that doesn't make sense to me as build-native.proj is a "Microsoft.Build.Traversal" project for which `UsingMicrosoftTraversalSdk` should be true and hence the IsSourceProject condition should already be false. Can you please check again just to be sure?

ViktorHofer replied:
Maybe it was native-binplace.proj?

sbomer committed:
Address feedback

- Don't reference ILLink.Tasks from subsets
- Use PackageReference from illink.tasks
- Import props from source directory

  Add extensibility to shipping targets to make this possible.
- Set IsSourceProject false for .proj files

ViktorHofer commented on eng/illink.targets:
```diff
@@ -1,5 +1,17 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="$(_ILLinkTasksSourceDir)ILLink.Tasks.csproj"
+                      ReferenceOutputAssembly="false" />
```
You need to add `PrivateAssets="all"` as otherwise packages would have a package dependency to the illink package.

ViktorHofer replied:
Also, make sure that the project is built in the same configuration as when doing a root build. By default, the referencing project's configuration flows to the ProjectReference project. We probably don't want that here.

ViktorHofer commented on src/libraries/oob.proj:
```diff
@@ -81,4 +78,12 @@
     <Message Condition="Exists('$(ArtifactsDir)packages')" Importance="High" Text="##vso[task.setvariable variable=_librariesBuildProducedPackages]true" />
   </Target>
 
+  <Import Project="Sdk.targets" Sdk="Microsoft.Build.NoTargets" />
+
+  <!-- Must be imported AFTER liveBuilds.targets (imported by Directory.Build.targets) which defines the ToolsILLinkDir. -->
+  <ImportGroup Condition="'$(BuildTargetFramework)' == '$(NetCoreAppCurrent)' or '$(BuildTargetFramework)' == ''">
+    <!-- Import the illink file which contains some of the logic required to illink the out-of-band assemblies. -->
+    <Import Project="$(RepositoryEngineeringDir)illink.targets" />
+  </ImportGroup>
```
nit: would you mind reverting these changes here and instead define another property in the Directory.Build.targets where illink.targets is imported: https://github.com/dotnet/runtime/blob/7cc9f149bf4cc481f96e4c2e856908fc49f4adf8/src/libraries/Directory.Build.targets#L130

`Condition="'$(IsSourceProject)' == 'true' or '$(ExplicitlyImportCustomILLinkTargets)' == 'true'"

sbomer committed:
Fix TFM, Configuration, PrivateAssets

sbomer committed:
Address feedback

Avoid needing to reorganize SDK imports

sbomer committed:
Fix typo

sbomer committed:
Fix configuration in ToolsILLinkDir

ViktorHofer commented on eng/illink.targets:
```diff
@@ -1,5 +1,23 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <!-- SkipGetTargetFrameworkProperties builds ILLink.Tasks without any TargetFramework settings
+         based on the referencing project, preventing errors when illink.targets is imported by a
+         project with TargetFrameworks that are incompatible with those of ILLink.Tasks. -->
+    <ProjectReference Include="$(_ILLinkTasksSourceDir)ILLink.Tasks.csproj"
+                      ReferenceOutputAssembly="false"
+                      PrivateAssets="all"
+                      SkipGetTargetFrameworkProperties="true"
+                      SetConfiguration="Configuration=$(ToolsConfiguration)" />
```
Nearly there. Now that you skip retrieving target framework properties, you need to pass the target framework in, explicitly: `SetTargetFramework="TargetFramework=$(NetCoreAppToolCurrent)"` or `$(NetFrameworkToolCurrent)`. And add a comment to keep these in sync with https://github.com/dotnet/runtime/blob/9c3f8b3727d9be4de483a1d725c2bda22f956688/src/tools/illink/src/ILLink.Tasks/ILLink.Tasks.csproj#L4

You want to use the .NETCoreApp one when msbuild runs on .NETCoreApp, otherwise use the .NETFramework TFM.

```suggestion
    <!-- SkipGetTargetFrameworkProperties builds ILLink.Tasks without any TargetFramework settings
         based on the referencing project, preventing errors when illink.targets is imported by a
         project with TargetFrameworks that are incompatible with those of ILLink.Tasks. -->
    <ProjectReference Include="$(_ILLinkTasksSourceDir)ILLink.Tasks.csproj"
                      ReferenceOutputAssembly="false"
                      PrivateAssets="all"
                      SkipGetTargetFrameworkProperties="true"
                      SetConfiguration="Configuration=$(ToolsConfiguration)">
        <!-- Keep TFMs in sync with ILLink.Tasks.csproj -->
        <SetTargetFramework Condition="'$(MSBuildRuntimeType)' == 'Core'">TargetFramework=$(NetCoreAppToolCurrent)</SetTargetFramework>
        <SetTargetFramework Condition="'$(MSBuildRuntimeType)' != 'Core'">TargetFramework=$(NetFrameworkToolCurrent)</SetTargetFramework>
    </ProjectReference>
```

sbomer committed:
Don't change IsSourceProject

sbomer committed:
Update eng/illink.targets

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

ViktorHofer committed:
Fix tools configuration

ViktorHofer commented:
@sbomer I pushed one more commit to your branch to make sure that the tools build with the correct configuration when building from root.

ViktorHofer committed:
Normalize directory separator chars

We agreed on (at least under src/libraries) to use the Windows directory separator char `\` instead of a forward slash. This helps when looking at binlogs as those then correctly display a path.

ViktorHofer commented:
Built your changes locally with a mixed set of configurations: `.\build.cmd clr+tools+libs -c Release -lc Debug /bl`. Looked at the binlog and didn't see any double writes introduced by this change.

lewing commented:
This looks like it may be breaking the official build https://dev.azure.com/dnceng/internal/_build/results?buildId=2257152&view=results

ViktorHofer commented:
Seems to be NativeAOT legs only.

ViktorHofer commented:
Fix is here: https://github.com/dotnet/runtime/pull/91454
